src/misc/coloring.c

```diff
@@ -129,6 +131,105 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
+            continue;
+        }
+        max_saturation = saturation;
+        max_degree = degree;
+        most_saturated_node = node;
+    }
+
+    return most_saturated_node;
+}
+
+static igraph_bool_t is_color_used_by_neighbour(
+    const igraph_vector_int_t *colors, igraph_integer_t color,
+    const igraph_vector_int_t *neighbours
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (VECTOR(*colors)[nbr] == color) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static void igraph_i_dsatur_update_saturation_degree(
+    const igraph_adjlist_t *adjlist, igraph_vector_int_t *saturation_degree,
+    igraph_vector_int_t *neighbours, igraph_vector_int_t *colors,
+    igraph_integer_t color
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++){
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (!is_color_used_by_neighbour(colors, color, igraph_adjlist_get(adjlist, nbr))) {
+            VECTOR(*saturation_degree)[nbr]++;
+        }
+    }
+}
+
+static igraph_error_t igraph_i_vertex_coloring_dsatur_cn(
+    const igraph_t *graph, igraph_vector_int_t *colors
+) {
+    igraph_adjlist_t adjlist;
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &adjlist, IGRAPH_ALL, IGRAPH_LOOPS_TWICE, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &adjlist);
+
+    igraph_vector_int_t saturation_degree;
+    IGRAPH_CHECK(igraph_vector_int_init(&saturation_degree, igraph_vcount(graph) ) );
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &saturation_degree);
+
+    igraph_vector_int_t edge_degree;
+    IGRAPH_CHECK(igraph_vector_int_init(&edge_degree, igraph_vcount(graph) ) );
+    IGRAPH_CHECK(igraph_degree(graph, &edge_degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS ) );
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &edge_degree);
+
+    igraph_vector_int_fill(colors, -1);   // -1 as a color means uncolored
+
+    igraph_integer_t vc = igraph_vcount(graph);
+    igraph_integer_t vertices_colored = 0;
+    while (vertices_colored < vc) {
+        igraph_integer_t node_to_color = igraph_i_dsatur_select_node(graph, colors, &saturation_degree, &edge_degree, vc);
+        igraph_vector_int_t *neighbours = igraph_adjlist_get(&adjlist, node_to_color);
+        for (igraph_integer_t color = 0; color < vc; color++) {
+            igraph_bool_t viable_color = !is_color_used_by_neighbour(colors, color, neighbours);
+            if (viable_color) {
+                igraph_i_dsatur_update_saturation_degree(&adjlist, &saturation_degree, neighbours, colors, color);
+                VECTOR(*colors)[node_to_color] = color;
+                break;
+            }
+        }
```

Comment:
Here you choose a new colour by iterating through all possible colours, and for each colour, iterating through neighbours of the current vertex. I believe it would be more efficient to retrieve neighbour colours once, then sort them. Iterate through the sorted list, and choose the lowest unused colour you find. This avoids several iterations through the neighbour list.